lib/Echidna/Mutator/Corpus.hs

```diff
@@ -0,0 +1,96 @@
+{-# LANGUAGE FlexibleContexts #-}
+
+module Echidna.Mutator.Corpus where
+
+import Control.Monad.Random.Strict (MonadRandom, getRandomR, weighted)
+import Control.Monad.State.Strict (MonadState(..))
+import Data.Has (Has(..))
+
+import qualified Data.Set as DS
+
+import Echidna.Types.Tx (Tx)
+import Echidna.Types.Corpus
+import Echidna.Transaction (mutateTx, shrinkTx)
+import Echidna.ABI (GenDict)
+import Echidna.Mutator.Array
+
+type MutationConsts = (Integer, Integer, Integer, Integer)
+defaultMutationConsts :: MutationConsts 
+defaultMutationConsts = (1,1,1,1)
+
+data TxsMutation = Shrinking
+                 | Mutation
+                 | Expansion
+                 | Swapping
+                 | Deletion
+
+data CorpusMutation = Skip
+                    | RandomAppend TxsMutation 
+                    | RandomPrepend TxsMutation
+                    | RandomSplice
+                    | RandomInterleave
+
+selectAndMutate :: MonadRandom m 
+                => ([Tx] -> m [Tx]) -> Corpus -> m [Tx]
+selectAndMutate f ctxs = do
+  rtxs <- weighted $ map (\(i, txs) -> (txs, fromInteger i)) $ DS.toDescList ctxs
+  k <- getRandomR (0, length rtxs - 1)
+  f $ take k rtxs
+
+selectAndCombine ::  MonadRandom m
+                 => ([Tx] -> [Tx] -> m [Tx]) -> Int -> Corpus -> [Tx] -> m [Tx]
+selectAndCombine f ql ctxs gtxs = do
+  rtxs1 <- selectFromCorpus
+  rtxs2 <- selectFromCorpus
+  txs <- f rtxs1 rtxs2
+  return . take ql $ txs ++ gtxs
+    where selectFromCorpus = weighted $ map (\(i, txs) -> (txs, fromInteger i)) $ DS.toDescList ctxs
+
+getCorpusMutation :: (MonadRandom m, Has GenDict x, MonadState x m) 
+                  => CorpusMutation -> (Int -> Corpus -> [Tx] -> m [Tx])
+getCorpusMutation Skip = \_ _ -> return
+getCorpusMutation (RandomAppend m) = 
+  case m of
+    Shrinking  -> mut (mapM shrinkTx)
+    Mutation   -> mut (mapM mutateTx)
+    Expansion  -> mut expandRandList
+    Swapping   -> mut swapRandList
+    Deletion   -> mut deleteRandList
+ where mut f ql ctxs gtxs = do
+          rtxs' <- selectAndMutate f ctxs
+          return . take ql $ rtxs' ++ gtxs  
+
```

Comment:
i don't think there should be newlines between function pattern matches

Reply:
Fixed